java/client/src/org/openqa/selenium/support/pagefactory/ByChained.java

```diff
@@ -48,10 +48,21 @@ public ByChained(By... bys) {
 
   @Override
   public WebElement findElement(SearchContext context) {
-    List<WebElement> elements = findElements(context);
-    if (elements.isEmpty())
-      throw new NoSuchElementException("Cannot locate an element using " + toString());
-    return elements.get(0);
+    if (bys.length == 0) {
+      throw new NoSuchElementException("No Bys were specified in this " + getClass().getSimpleName());
+    }
+
+    WebElement elem = null;
+    for (By by : bys) {
+      boolean isFirstIteration = elem == null;
+      if (isFirstIteration) {
+        elem = by.findElement(context);
+      } else {
+        elem = elem.findElement(by);
+      }
+    }
+    assert elem != null; // NoSuchElement should have already been thrown
```

Comment:
Let's not use "assert" but rather throw an Exception explicitly.

Overall, this logic re-write doesn't maintain the existing functionality. Namely if any of the by's do not find the element, it doesn't throw an Element Not Found exception, but rather continues executing all by's until at least one was found.

I would change your for loop to look more like this:

    for (By by : bys) {
      List<WebElements> elements = by.findElements(elem == null ? context : elem);
      if (!elements.isEmpty()) {
        elem = elements.get(0);
      }
    }